UOP1_Project/Assets/Scripts/Characters/Character.cs

```diff
@@ -29,69 +29,112 @@ public class Character : MonoBehaviour
 	private bool shouldSlide; // Should player slide?
 	private Vector3 inputVector; //Initial input horizontal movement (y == 0f)
 	private Vector3 movementVector; //Final movement vector
+	private StateMachine stateMachine; // handles different states and transitions between them
 
 	private const float ROTATION_TRESHOLD = .02f; // Used to prevent NaN result causing rotation in a non direction
 
 	private void Awake()
 	{
 		characterController = GetComponent<CharacterController>();
+		OrchestrateStateMachine();
 	}
 
 	private void Update()
 	{
-		//Raises the multiplier to how much gravity will affect vertical movement when in mid-air
-		//This is 0f at the beginning of a jump and will raise to maximum 1f
-		if (!characterController.isGrounded)
-		{
-			gravityContributionMultiplier += Time.deltaTime * gravityComebackMultiplier;
-		}
-		//Reduce the influence of the gravity while holding the Jump button
-		if (isJumping)
+		// calls tick on the currently active state
+		stateMachine.Tick();
+	}
+
+	private void OrchestrateStateMachine()
+	{
+		stateMachine = new StateMachine();
+
+		// create all states and pass in references to this character
+		IdleState idleState = new IdleState(this);
+		WalkingState walkingState = new WalkingState(this);
+		JumpingState jumpingState = new JumpingState(this);
+		FallingState fallingState = new FallingState(this);
+
+		// setup all explicit transitions
+		idleState.AddTransition(jumpingState, PerformedJumpAction());
+		idleState.AddTransition(walkingState, IsMovingAndGrounded());
+		walkingState.AddTransition(jumpingState, PerformedJumpAction());
+		fallingState.AddTransition(idleState, IsCharacterNotMovingAndGrounded());
```

Comment:
I believe there is a redundacy with l.66 
(stateMachine.AddAnyTransition(idleState, IsCharacterNotMovingAndGrounded());)